### ISSUE TITLE
Use hyperdiscovery

### ISSUE THREAD

**Issue by RangerMauve** [cabal-club/cabal-core#44]
What do y'all think about adopting [hyperdiscovery](https://github.com/datproject/hyperdiscovery) for the swarm rather than using the raw discovery-swarm commands?

This'd be a breaking change since it'd use a different discovery key and would use the raw hypercore protocol for handshaking rather than the default discovery-swarm handshake.

So, breaking changes are scary, but this could be very beneficial: It'd make it easy to run on the web using discovery-swarm-web (getting that running has been a huge PITA). It'd also be less maintenance since we could reuse code between the rest of the dat ecosystem. Finally, once we get hyperswarm integrated into hyperdiscovery, it should be a minor version bump to get all it's benefits into Cabal. 😁

Would a PR introducing this breaking change be welcome?

**Comment by cblgh** [comment]
@RangerMauve hell yeah bring on the breakage

(sorry for inadvertently making your life hard last year by not using hyperdiscovery to start with!!!)

**Comment by RangerMauve** [comment]
I don't think anyone has really been using hyperdiscovery yet, and you were using discovery-swarm according to it's defaults which makes sense. It's just that the dat CLI and Beaker weren't using discovery-swarm's defaults. :P

**Comment by karissa** [comment]
This would also benefit cabal-desktop so multiple cabals would share the same port, making it easier to push past firewalls in case you need to open a port for it

**Comment by noffle** [comment]
Break us! :hammer: :boom: 

Some notes:
- How well tested is hyperdiscovery? Has it been deployed on dat yet?
- Does this mean dropping discovery-swarm, or just the default handshake? We'll retain local-net discovery right?
- If we want to retain the ability to have peers tell each other, upon connecting, who they are (pubkey), we'll need to add some kind of handshake. Doing this [is already on my list](https://github.com/cabal-club/cabal-core/issues/40) and is *also* a breaking change.

tl;dr: Upgrading to hyperdiscovery seems like a good excuse to do the breaking change I wanted to do too. Happy to coordinate this work /w you @RangerMauve!

**Comment by RangerMauve** [comment]
- hyperdiscovery hasn't been deployed in the CLI yet, but I'm using it in dat-store and it's working pretty well
- Hyperdiscovery is just using discovery-swarm under the hood with some defaults that were used in the dat ecosystem like dropping the default handshake and sha1. Local discovery should work just like before.
- I'd suggest doing the pubkey exchange through the [session data extension](https://www.datprotocol.com/deps/0006-session-data-extension/) so that it's still the raw hypercore protocol.

I could work on it next week. Here's what needs to be done:

- Update [swarm.js](https://github.com/cabal-club/cabal-core/blob/master/swarm.js) to use the latest hyperdiscovery
- Update the CLI / Desktop / ETC with the new changes
- Hopefully it'll work. :P

**Comment by noffle** [comment]
@RangerMauve does this mean we need to make sure we hash the cabal key ourselves? It'd be nice to expose a property like `discoveryKey` on cabal-core.

**Comment by RangerMauve** [comment]
Yeah, I think having a `discoveryKey` property in cabal-core would be really handy.

That's related to [this proposal](https://github.com/datprotocol/DEPs/issues/60) @Frando had for an `abstract-dat-interface` which different dat-based modules could adhere to so we can more easily plug them into shared code.

**Comment by RangerMauve** [comment]
So here's the interface that cabal needs to expose in order to use it with hyperdiscovery:

https://github.com/datproject/hyperdiscovery/blob/master/index.js#L189

- [ ] `cabal.ready(fn)`
- [ ]`cabal.discoveryKey` which needs to be 32 characters [due to it being parsed by dat-encoding (might be a bug)](https://github.com/datproject/hyperdiscovery/blob/master/index.js#L192)
- [ ] `cabal.key` the public key
- [ ] `cabal.replicate()`
- [ ] `cabal.peers`
- [ ] `cabal.isSwarming`

**Comment by karissa** [comment]
hyperdiscovery probably doesn't need to pass it through dat-encoding, although having a long discovery key can be useful for entropy

**Comment by RangerMauve** [comment]
Yeah, I think that part should be removed. Especially since it's not using dat-encoding for the key just a few lines later. :P

**Comment by RangerMauve** [comment]
Also, I think we'll need to use the regular hypercore discovery key after all. :/

Hyperdiscovery will let us reuse a swarm across cabals, but determining which incoming connections are coming from which cabal requires [the hyper core protocol's discovery keys](https://github.com/datproject/hyperdiscovery/blob/master/index.js#L136) which I think would be based on the main hypercore for a cabal.

Any objections to this? It'd be a breaking change regardless.

**Comment by cblgh** [comment]
@RangerMauve hm that might be an issue as we don't have a main hypercore, and instead just a [public key](https://github.com/cabal-club/cabal/blob/master/cli.js#L203) that we find each other over (same kind of public key as a hypercore tho). will that be a problem?

looks like it wouldn't be an issue if we just toss in a discovery key version of the cabal key, at least going by the [dat-encoding docs](https://github.com/dat-land/dat-encoding#tostrbuf) - but i might be missing something

**Comment by RangerMauve** [comment]
What's the first `feed` message that gets sent down the wire. Is it not related to the cabal key?

**Comment by cblgh** [comment]
@RangerMauve i think it might be the [bootstrapping hypercore](https://github.com/kappa-db/multifeed/blob/master/index.js#L38) that multifeed uses, [more info](https://github.com/kappa-db/multifeed#hacks). @noffle would know more

**Comment by RangerMauve** [comment]
So, the first feed is using the [hardcoded multifeed](https://github.com/kappa-db/multifeed/blob/master/index.js#L41) hypercore.

What if instead of using the hardcoded key, we used the cabal key?
That would improve the security a little since at the moment they could discovery the discovery key on the network and connect to the group without needing the original cabal key.

Plus it'd differentiate which incoming connection is coming from where and allow us to reuse swarms between cabals.

**Comment by cblgh** [comment]
hyperdiscovery says its deprecated
> DEPRECATED: see hyperswarm/replicator for similar functionality with hypercore
